app/services/oauth_clients/connection_manager.rb

```diff
@@ -0,0 +1,214 @@
+      client = rack_oauth_client # Configure and start the rack-oauth2 client
+      client.authorization_uri(scope:, state:)
+    end
+
+    # For the OAuth2 callback page: Calculate the redirection URL that will
+    # point the browser at the initial page that wanted to access the OAuth2
+    # protected resource.
+    # @param state (OAuth2 RFC) encapsulates the state of the calling page (URL + params) to return
+    def callback_redirect_uri(state)
+      # In the current implementation "state" just consists of the URL of
+      # the initial page, possibly with "&var=value" added parameters.
+      # So we can just return this URI.
+      state
+    end
+
+    # Called by callback_page with a cryptographic "code" that indicates
+    # that the user has successfully authorized the OAuth2 Authorization Server.
+    # We now are going to exchange this code to a token (bearer+refresh)
+    def code_to_token(code)
+      # Return a Rack::OAuth2::AccessToken::Bearer or an error string
+      service_result = request_new_token(authorization_code: code)
+      return service_result unless service_result.success?
+
+      # Create a new OAuthClientToken from Rack::OAuth::AccessToken::Bearer and return
+      ServiceResult.new(
+        success: true,
+        result: create_new_oauth_client_token(service_result.result)
+      )
+    end
+
+    private
+
+    # Check if a OAuthClientToken already exists and return nil otherwise.
+    # Don't handle the case of an expired token.
+    def get_existing_token
+      # Check if we've got a token in the database and return nil otherwise.
+      OAuthClientToken.find_by(user_id: @user, oauth_client_id: @oauth_client.id)
+    end
+
+    # Calls client.access_token!
+    # Convert the various exceptions into user-friendly error strings.
+    def request_new_token(options = {})
+      rack_access_token = rack_oauth_client(options)
+                            .access_token!(:body) # Rack::OAuth2::AccessToken
+
+      ServiceResult.new(success: true,
+                        result: rack_access_token)
+    rescue Rack::OAuth2::Client::Error => e # Handle Rack::OAuth2 specific errors
+      service_result_with_error(i18n_rack_oauth2_error_message(e))
+    rescue Timeout::Error, EOFError, Net::HTTPBadResponse, Net::HTTPHeaderSyntaxError, Net::ProtocolError,
+           Errno::EINVAL, Errno::ENETUNREACH, Errno::ECONNRESET, Errno::ECONNREFUSED, JSON::ParserError => e
+      service_result_with_error(
+        "#{I18n.t('oauth_client.errors.oauth_returned_http_error')}: #{e.class}: #{e.message.to_html}"
+      )
+    rescue StandardError => e
+      service_result_with_error(
+        "#{I18n.t('oauth_client.errors.oauth_returned_standard_error')}: #{e.class}: #{e.message.to_html}"
+      )
+    end
+
+    # Localize the error message
+    def i18n_rack_oauth2_error_message(rack_oauth2_client_exception)
+      l10n_key = "oauth_client.errors.rack_oauth2.#{rack_oauth2_client_exception.message}"
```

Comment:
why is the variable named `l10n_key`? not `i18n_key`? we do use internationalization, not only localization. And the key is constructed on our side.